tests/src/test/java/org/eclipse/hono/tests/client/ClientTestBase.java

```diff
@@ -230,7 +201,58 @@ private void disconnect(final TestContext ctx) {
      * @param messageConsumer The handler to invoke for every message received.
      * @return A future succeeding with the created consumer.
      */
-    abstract Future<MessageConsumer> createConsumer(String tenantId, Consumer<Message> messageConsumer);
+    protected abstract Future<MessageConsumer> createConsumer(String tenantId, Consumer<Message> messageConsumer);
+
+    /**
+     * Connects the downstream consumer client for connection to the messaging network.
+     * 
+     * @return The client for connecting to the messaging network.
+     */
+    protected abstract Future<HonoClient> connectDownstreamClient(ProtonClientOptions options);
+
+    /**
+     * Connects the AMQP 1.0 client to one of Hono's services.
+     * 
+     * @param options The options to use for the connection.
+     * 
+     * @return The client for connecting to a Hono service.
+     */
+    protected abstract Future<HonoClient> connectToService(ProtonClientOptions options, TestContext ctx);
+
+    /**
+     * Sends a message to the downstream consumer through a Hono service.
+     * 
+     * @param deviceId The id of the device sending the message.
+     * @param ctx The Vert.x test context.
+     */
+    protected abstract void sendMessage(String payload, TestContext ctx);
+
+    /**
+     * The id of the tenant under test.
+     * 
+     * @return The tenant identifier created during the test.
+     */
+    protected abstract String getTenantId();
```

Comment:
why do you want to store this in a class level field? I would assume that the tenant(s) and device(s) required by the test cases are created as part of the test cases themselves, using the helper to get random identifiers ...